src/ast.rs

```diff
@@ -0,0 +1,291 @@
+use token;
+use std::fmt;
+use std::collections::HashMap;
+use std::hash::{Hash, Hasher};
+
+#[derive(Debug)]
+pub enum Node {
+    Program(Box<Program>),
+    Statement(Box<Statement>),
+    Expression(Box<Expression>),
+}
+
+#[derive(Hash, Eq, PartialEq, Clone, Debug)]
+pub enum Statement {
+    Let(Box<LetStatement>),
+    Return(Box<ReturnStatement>),
+    Expression(Box<ExpressionStatement>),
+}
+
+impl fmt::Display for Statement {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        let s = match self {
+            Statement::Let(stmt) => format!("{}", stmt),
+            Statement::Return(ret) => format!("{}", ret),
+            Statement::Expression(exp) => format!("{}", exp),
+        };
+        write!(f, "{}", s)
+    }
+}
+
+#[derive(Debug, PartialEq, Eq, Hash, Clone)]
+pub enum Expression {
+    Identifier(String),
+    Integer(i64),
+    Prefix(Box<PrefixExpression>),
+    Infix(Box<InfixExpression>),
+    Boolean(bool),
+    String(String),
+    If(Box<IfExpression>),
+    Function(Box<FunctionLiteral>),
+    Call(Box<CallExpression>),
+    Array(Box<ArrayLiteral>),
+    Index(Box<IndexExpression>),
+    Hash(Box<HashLiteral>),
+}
+
+impl fmt::Display for Expression {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        let s = match self {
+            Expression::Identifier(s) => s.clone(),
+            Expression::Integer(value) => format!("{}", value),
+            Expression::Prefix(pref) => pref.to_string(),
+            Expression::Infix(infix) => infix.to_string(),
+            Expression::Boolean(b) => b.to_string(),
+            Expression::String(s) => s.clone(),
+            Expression::If(exp) => exp.to_string(),
+            Expression::Function(f) => f.to_string(),
+            Expression::Call(c) => c.to_string(),
+            Expression::Array(a) => a.to_string(),
+            Expression::Index(i) => i.to_string(),
+            Expression::Hash(h) => h.to_string(),
+        };
+        write!(f, "{}", s)
+    }
+}
+
+#[derive(Eq, PartialEq, Clone, Debug)]
+pub struct HashLiteral {
+    pub pairs: HashMap<Expression,Expression>,
+}
+
+// Had to implement Hash for this because HashMap doesn't. Doesn't matter what this is because
+// a HashLiteral isn't a valid expression as a key in a monkey hash.
+impl Hash for HashLiteral {
+    fn hash<H: Hasher>(&self, _state: &mut H) {
+        panic!("hash not implemented for HashLiteral");
+    }
+}
+
+impl fmt::Display for HashLiteral {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        let pairs: Vec<String> = (&self.pairs).into_iter().map(|(k, v)| format!("{}:{}", k.to_string(), v.to_string())).collect();
+        write!(f, "{{{}}}", pairs.join(", "))
+    }
+}
+
+#[derive(Hash, Eq, PartialEq, Clone, Debug)]
+pub struct IfExpression {
+    pub condition: Expression,
+    pub consequence: BlockStatement,
+    pub alternative: Option<BlockStatement>,
```

Comment:
You can use `Option` for any type that has a fixed size (that is `Sized`), e.g. structs, enums. `Box` is usually used for trait objects (as in `Box<dyn Trait>` when the trait is "object-safe" (not all traits can be turned into trait objects)) or sometimes boxed slices (`Box<[T]>` because while a reference to a slice is Sized (it's a fat pointer that also stores the len), a slice itself is not Sized.
Btw, a reference to a trait object is also a fat pointer (pointer to the instance and a pointer to the vtable of its trait impl).
The doc has a detailed explanation of the rules for when a trait is object-safe (sorry, I'm on my phone right now).

Box of course also differs from Option in that it represents a definitely existing T instead of an optional T, and thus automatically dereferences to its content (via the `Deref` trait).

Btw, I just read your blog post and I'm interested in helping you on your journey of learning Rust and using it in production :) 

You mentioned that you're also interested in gRPC and web servers. There are several crates for gRPC (when I tried it there was only one but it was working well) and for Web servers I've been using rocket in production (before actix-web existed) and now also actix-web (because it's async (and it allows serving websockets on the same port)).

Reply:
Another +1 on actix/actix-web, I'm surprised at just how solid it is and it seems to be taking over in any place I see such programming styles being used.

Reply:
@Boscop thanks, that's great info. I'll have to read more up on trait objects. Didn't really get a chance to utilize traits in this code, but it's coming up for some future hacking. I'll search through the gRPC crates. I saw in some thread criticism about Rocket because it's not async so I'll probably dig into actix first.